<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Hongyu Guo" />

<meta name="date" content="2021-03-16" />

<title>Single-cell Sorter</title>

<link href="data:text/css;charset=utf-8,%0Aa%2Eanchor%2Dsection%20%7Bmargin%2Dleft%3A%2010px%3B%20visibility%3A%20hidden%3B%20color%3A%20inherit%3B%7D%0Aa%2Eanchor%2Dsection%3A%3Abefore%20%7Bcontent%3A%20%27%23%27%3B%7D%0A%2EhasAnchor%3Ahover%20a%2Eanchor%2Dsection%20%7Bvisibility%3A%20visible%3B%7D%0A" rel="stylesheet" />
<script src="data:application/x-javascript;base64,Ly8gQW5jaG9yIHNlY3Rpb25zIHYxLjAgd3JpdHRlbiBieSBBdHN1c2hpIFlhc3Vtb3RvIG9uIE9jdCAzcmQsIDIwMjAuCmRvY3VtZW50LmFkZEV2ZW50TGlzdGVuZXIoJ0RPTUNvbnRlbnRMb2FkZWQnLCBmdW5jdGlvbigpIHsKICAvLyBEbyBub3RoaW5nIGlmIEFuY2hvckpTIGlzIHVzZWQKICBpZiAodHlwZW9mIHdpbmRvdy5hbmNob3JzID09PSAnb2JqZWN0JyAmJiBhbmNob3JzLmhhc093blByb3BlcnR5KCdoYXNBbmNob3JKU0xpbmsnKSkgewogICAgcmV0dXJuOwogIH0KCiAgY29uc3QgaCA9IGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3JBbGwoJ2gxLCBoMiwgaDMsIGg0LCBoNSwgaDYnKTsKCiAgLy8gRG8gbm90aGluZyBpZiBzZWN0aW9ucyBhcmUgYWxyZWFkeSBhbmNob3JlZAogIGlmIChBcnJheS5mcm9tKGgpLnNvbWUoeCA9PiB4LmNsYXNzTGlzdC5jb250YWlucygnaGFzQW5jaG9yJykpKSB7CiAgICByZXR1cm4gbnVsbDsKICB9CgogIC8vIFVzZSBzZWN0aW9uIGlkIHdoZW4gcGFuZG9jIHJ1bnMgd2l0aCAtLXNlY3Rpb24tZGl2cwogIGNvbnN0IHNlY3Rpb25faWQgPSBmdW5jdGlvbih4KSB7CiAgICByZXR1cm4gKCh4LmNsYXNzTGlzdC5jb250YWlucygnc2VjdGlvbicpIHx8ICh4LnRhZ05hbWUgPT09ICdTRUNUSU9OJykpCiAgICAgICAgICAgID8geC5pZCA6ICcnKTsKICB9OwoKICAvLyBBZGQgYW5jaG9ycwogIGguZm9yRWFjaChmdW5jdGlvbih4KSB7CiAgICBjb25zdCBpZCA9IHguaWQgfHwgc2VjdGlvbl9pZCh4LnBhcmVudEVsZW1lbnQpOwogICAgaWYgKGlkID09PSAnJykgewogICAgICByZXR1cm4gbnVsbDsKICAgIH0KICAgIGxldCBhbmNob3IgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdhJyk7CiAgICBhbmNob3IuaHJlZiA9ICcjJyArIGlkOwogICAgYW5jaG9yLmNsYXNzTGlzdCA9IFsnYW5jaG9yLXNlY3Rpb24nXTsKICAgIHguY2xhc3NMaXN0LmFkZCgnaGFzQW5jaG9yJyk7CiAgICB4LmFwcGVuZENoaWxkKGFuY2hvcik7CiAgfSk7Cn0pOwo="></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Single-cell Sorter</h1>
<h4 class="author">Hongyu Guo</h4>
<h4 class="date">2021-03-16</h4>



<p>The scSorter package implements the semi-supervised cell type assignment algorithm described in &quot;scSorter: assigning cells to known cell types according to known marker genes&quot;. This algorithm assigns cells to known cell types, assuming that the identities of marker genes are given but the exact expression levels of marker genes are unavailable. This vignette will illustrate the steps to apply the algorithm to single-cell RNA sequencing (scRNA-seq) data.</p>
<div id="section-1---preliminaries" class="section level2">
<h2>Section 1 - Preliminaries</h2>
<p>scSorter takes as input data the expression matrix from single-cell RNA sequencing and the annotation file that specifies the names of marker genes for each cell type of interest. scSorter assumes the input expression data have been properly normalized for the library size and have been properly transformed (e.g. log-transformation) to stabilize the variance.</p>
<p>To start the analysis, load scSorter package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scSorter)</code></pre></div>
<div id="real-data-example---tm-pancreas" class="section level3">
<h3>Real Data Example - TM pancreas</h3>
<p>In this vignette, we will use the data created by the Tabula Muris Consortium <span class="citation">(Tabula Muris Consortium and others 2018)</span> to illustrate the procedure for scSorter to assign cells to known cell types. We follow the analysis presented in our paper and focus on pancreas tissue from which 1,564 cells with valid cell type annotation are available. The cells are from: pancreatic A (390 cells), pancreatic B (449 cells), pancreatic D (140 cells), pancreatic PP (73 cells), pancreatic acinar (182 cells), pancreatic ductal (161 cells), pancreatic stellate (49 cells), endothelial (66 cells) and immune (54 cells). Marker genes for these cell types are extracted from the original study.</p>
</div>
<div id="examing-the-data" class="section level3">
<h3>Examing the data</h3>
<p>We first load the data to ensure they are in the correct formats.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="kw">url</span>(<span class="st">'https://github.com/hyguo2/scSorter/blob/master/inst/extdata/TMpancreas.RData?raw=true'</span>))</code></pre></div>
<p>The expression matrix <em>expr</em> should represent genes by rows and cells by columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;               A1.MAA000884.3_10_M.1.1 A1.MAA001862.3_39_F.1.1</span>
<span class="co">#&gt; 0610005C13Rik                       0                       0</span>
<span class="co">#&gt; 0610007C21Rik                     168                     162</span>
<span class="co">#&gt; 0610007L01Rik                       1                       2</span>
<span class="co">#&gt; 0610007N19Rik                       0                       0</span>
<span class="co">#&gt; 0610007P08Rik                       0                       0</span>
<span class="co">#&gt;               A10.MAA000884.3_10_M.1.1 A10.MAA001857.3_38_F.1.1</span>
<span class="co">#&gt; 0610005C13Rik                        0                        0</span>
<span class="co">#&gt; 0610007C21Rik                      121                       18</span>
<span class="co">#&gt; 0610007L01Rik                       44                      170</span>
<span class="co">#&gt; 0610007N19Rik                        7                       22</span>
<span class="co">#&gt; 0610007P08Rik                        2                        0</span>
<span class="co">#&gt;               A11.MAA000577.3_8_M.1.1</span>
<span class="co">#&gt; 0610005C13Rik                       0</span>
<span class="co">#&gt; 0610007C21Rik                     373</span>
<span class="co">#&gt; 0610007L01Rik                       0</span>
<span class="co">#&gt; 0610007N19Rik                       0</span>
<span class="co">#&gt; 0610007P08Rik                       0</span>
<span class="kw">dim</span>(expr)
<span class="co">#&gt; [1] 23433  1564</span></code></pre></div>
<p>Next, we check the annotation file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(anno)
<span class="co">#&gt;                 Type Marker Weight</span>
<span class="co">#&gt; 1  Endothelial_cells Pecam1      2</span>
<span class="co">#&gt; 2  Endothelial_cells   Cdh5      2</span>
<span class="co">#&gt; 3  Endothelial_cells    Kdr      2</span>
<span class="co">#&gt; 4             Immune  Ptprc      2</span>
<span class="co">#&gt; 5 Pancreatic_A_cells    Gcg      2</span>
<span class="co">#&gt; 6 Pancreatic_A_cells   Mafb      2</span></code></pre></div>
<p>The annotation file could be stored in a matrix or a data frame. Two columns, &quot;Type&quot; and &quot;Marker&quot;, are mandatory as they record the names of marker genes for each cell type of interest and will guide scSorter to assign cells to each cell type. The third column &quot;Weight&quot; is optional. As explained in our paper, weights could be assigned to each marker gene to represent their relative importance during the cell type assignment. A larger weight reflects more confidence of a gene being a marker gene of the corresponding cell type. If such knowledge is not available, we recommend simply choose a constant value for all marker genes by setting the <em>default_weight</em> option in the scSorter main function and the &quot;Weight&quot; column could therefore be omitted.</p>
<p>Now that the formats of these data are correct, we could proceed to the next step.</p>
</div>
</div>
<div id="section-2---preprocessing-the-data" class="section level2">
<h2>Section 2 - Preprocessing the data</h2>
<p>scSorter does not need all genes from the original expression matrix to conduct cell type assignment. Instead, it uses only marker genes and a selected group of highly variable genes other than marker genes to conduct the analysis. scSorter also requires the input expression data to be properly normalized for the library size and properly transformed to stabilize the variance. Here is a simple example of how this can be done:</p>
<p>Choose highly variable genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes =<span class="st"> </span><span class="kw">xfindvariable_genes</span>(expr, <span class="dt">ngenes =</span> <span class="dv">2000</span>)</code></pre></div>
<p>Normalize the input data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr =<span class="st"> </span><span class="kw">xnormalize_scData</span>(expr)</code></pre></div>
<p>We also filter out genes with non-zero expression in less than 10% of total cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgene_filter =<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">as.matrix</span>(expr)[topgenes, ]<span class="op">!=</span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="kw">ncol</span>(expr)<span class="op">*</span>.<span class="dv">1</span>
topgenes =<span class="st"> </span>topgenes[topgene_filter]</code></pre></div>
<p>At last, we subset the preprocessed expression data. Now, we are ready to run scSorter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">picked_genes =<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(anno<span class="op">$</span>Marker, topgenes))
expr =<span class="st"> </span>expr[<span class="kw">rownames</span>(expr) <span class="op">%in%</span><span class="st"> </span>picked_genes, ]</code></pre></div>
<p>The above code is only for showing the concept. In practice, there are plenty of packages that could preprocess the data. Here, we use the package Seurat to illustrate this step (please remove the # mark in the code when you run it).</p>
<p>First, we normalize and transform the data with NormalizeData() function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(Seurat)</span>
<span class="co">#expr_obj = CreateSeuratObject(expr)</span>
<span class="co">#expr_obj &lt;- NormalizeData(expr_obj, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000, verbose = F)</span></code></pre></div>
<p>Then, we choose highly variable genes by FindVariableFeatures() function. We also filter out genes with non-zero expression in less than 10% of total cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#expr_obj &lt;- FindVariableFeatures(expr_obj, selection.method = &quot;vst&quot;, nfeatures = 2000, verbose = F)</span>
<span class="co">#topgenes &lt;- head(VariableFeatures(expr_obj), 2000)</span>

<span class="co">#expr = GetAssayData(expr_obj)</span>
<span class="co">#topgene_filter = rowSums(as.matrix(expr)[topgenes, ]!=0) &gt; ncol(expr)*.1</span>
<span class="co">#topgenes = topgenes[topgene_filter]</span></code></pre></div>
<p>At last, we subset the preprocessed expression data. Now, we are ready to run scSorter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#picked_genes = unique(c(anno$Marker, topgenes))</span>
<span class="co">#expr = expr[rownames(expr) %in% picked_genes, ]</span></code></pre></div>
</div>
<div id="section-3---running-scsorter" class="section level2">
<h2>Section 3 - Running scSorter</h2>
<p>The scSorter function requires the preprocessed data as input. As mentioned in Section 1, if the &quot;Weight&quot; column is omitted in the annotation file, a single weight should be specified for marker genes by using the <em>default_weight</em> option. Otherwise, the algorithm will use the weights from the annotation file to conduct the analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rts &lt;-<span class="st"> </span><span class="kw">scSorter</span>(expr, anno)</code></pre></div>
<div id="viewing-results" class="section level3">
<h3>Viewing Results</h3>
<p>The cell type assignment results are stored in the Pred_Type vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">table</span>(rts<span class="op">$</span>Pred_Type))
<span class="co">#&gt; </span>
<span class="co">#&gt;         Endothelial_cells                    Immune        Pancreatic_A_cells </span>
<span class="co">#&gt;                        63                        41                       416 </span>
<span class="co">#&gt;   Pancreatic_Acinar_cells        Pancreatic_B_cells        Pancreatic_D_cells </span>
<span class="co">#&gt;                       185                       457                       116 </span>
<span class="co">#&gt;   Pancreatic_Ductal_cells       Pancreatic_PP_cells Pancreatic_Stellate_cells </span>
<span class="co">#&gt;                       162                        73                        33 </span>
<span class="co">#&gt;                   Unknown </span>
<span class="co">#&gt;                        18</span></code></pre></div>
<p>The misclassification rate is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mis_rate =<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(rts<span class="op">$</span>Pred_Type <span class="op">==</span><span class="st"> </span>true_type)
<span class="kw">round</span>(mis_rate, <span class="dv">4</span>)
<span class="co">#&gt; [1] 0.055</span></code></pre></div>
<p>The confusion matrix is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(true_type, rts<span class="op">$</span>Pred_Type)
<span class="co">#&gt;                            </span>
<span class="co">#&gt; true_type                   Endothelial_cells Immune Pancreatic_A_cells</span>
<span class="co">#&gt;   Endothelial_cells                        63      0                  0</span>
<span class="co">#&gt;   Immune                                    0     41                  6</span>
<span class="co">#&gt;   Pancreatic_A_cells                        0      0                384</span>
<span class="co">#&gt;   Pancreatic_Acinar_cells                   0      0                  0</span>
<span class="co">#&gt;   Pancreatic_B_cells                        0      0                  0</span>
<span class="co">#&gt;   Pancreatic_D_cells                        0      0                 16</span>
<span class="co">#&gt;   Pancreatic_Ductal_cells                   0      0                  0</span>
<span class="co">#&gt;   Pancreatic_PP_cells                       0      0                  9</span>
<span class="co">#&gt;   Pancreatic_Stellate_cells                 0      0                  1</span>
<span class="co">#&gt;                            </span>
<span class="co">#&gt; true_type                   Pancreatic_Acinar_cells Pancreatic_B_cells</span>
<span class="co">#&gt;   Endothelial_cells                               2                  1</span>
<span class="co">#&gt;   Immune                                          3                  0</span>
<span class="co">#&gt;   Pancreatic_A_cells                              0                  1</span>
<span class="co">#&gt;   Pancreatic_Acinar_cells                       176                  2</span>
<span class="co">#&gt;   Pancreatic_B_cells                              0                446</span>
<span class="co">#&gt;   Pancreatic_D_cells                              0                  6</span>
<span class="co">#&gt;   Pancreatic_Ductal_cells                         1                  0</span>
<span class="co">#&gt;   Pancreatic_PP_cells                             0                  1</span>
<span class="co">#&gt;   Pancreatic_Stellate_cells                       3                  0</span>
<span class="co">#&gt;                            </span>
<span class="co">#&gt; true_type                   Pancreatic_D_cells Pancreatic_Ductal_cells</span>
<span class="co">#&gt;   Endothelial_cells                          0                       0</span>
<span class="co">#&gt;   Immune                                     0                       0</span>
<span class="co">#&gt;   Pancreatic_A_cells                         0                       0</span>
<span class="co">#&gt;   Pancreatic_Acinar_cells                    0                       0</span>
<span class="co">#&gt;   Pancreatic_B_cells                         0                       0</span>
<span class="co">#&gt;   Pancreatic_D_cells                       115                       0</span>
<span class="co">#&gt;   Pancreatic_Ductal_cells                    1                     159</span>
<span class="co">#&gt;   Pancreatic_PP_cells                        0                       0</span>
<span class="co">#&gt;   Pancreatic_Stellate_cells                  0                       3</span>
<span class="co">#&gt;                            </span>
<span class="co">#&gt; true_type                   Pancreatic_PP_cells Pancreatic_Stellate_cells</span>
<span class="co">#&gt;   Endothelial_cells                           0                         0</span>
<span class="co">#&gt;   Immune                                      0                         0</span>
<span class="co">#&gt;   Pancreatic_A_cells                          5                         0</span>
<span class="co">#&gt;   Pancreatic_Acinar_cells                     1                         0</span>
<span class="co">#&gt;   Pancreatic_B_cells                          3                         0</span>
<span class="co">#&gt;   Pancreatic_D_cells                          3                         0</span>
<span class="co">#&gt;   Pancreatic_Ductal_cells                     0                         0</span>
<span class="co">#&gt;   Pancreatic_PP_cells                        61                         0</span>
<span class="co">#&gt;   Pancreatic_Stellate_cells                   0                        33</span>
<span class="co">#&gt;                            </span>
<span class="co">#&gt; true_type                   Unknown</span>
<span class="co">#&gt;   Endothelial_cells               0</span>
<span class="co">#&gt;   Immune                          4</span>
<span class="co">#&gt;   Pancreatic_A_cells              0</span>
<span class="co">#&gt;   Pancreatic_Acinar_cells         3</span>
<span class="co">#&gt;   Pancreatic_B_cells              0</span>
<span class="co">#&gt;   Pancreatic_D_cells              0</span>
<span class="co">#&gt;   Pancreatic_Ductal_cells         0</span>
<span class="co">#&gt;   Pancreatic_PP_cells             2</span>
<span class="co">#&gt;   Pancreatic_Stellate_cells       9</span></code></pre></div>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-tmpancreas">
<p>Tabula Muris Consortium and others. 2018. “Single-Cell Transcriptomics of 20 Mouse Organs Creates a Tabula Muris.” <em>Nature</em> 562 (October). Genome Research: 367–72. doi:<a href="https://doi.org/10.1038/s41586-018-0590-4">10.1038/s41586-018-0590-4</a>.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
